Add 5th denom, drop deprecated logic and bump min PS version

UdjinM6 opened the pull request:
- add 5th denom (0.001)
- always allow OP_RETURN in PS collaterals (i.e. drop deprecated logic)
- collateral amount is always 1/10 of the smallest denom (i.e. no need to update it manually if we are going to add yet another denom later)
- bump MIN_PRIVATESEND_PEER_PROTO_VERSION to ~70211~70212 to avoid conflicts with 70210 (esp. during migration)
- drop backward compatibility in dsa/dsq serialization (we won't process messages from nodes below MIN_PRIVATESEND_PEER_PROTO_VERSION anyway)

Note: After merging this `develop` won't be able to mix on current mainnet until migration, you'd have to revert it locally to keep mixing on mainnet.

PastaPastaPasta commented:
How will the rollout occur? Will some users not be able to mix if they have the new version and the majority of the network doesn't? Based on what you said the upgrade doesn't sound like it would be smooth

PastaPastaPasta commented:
Also, mostly fixes https://github.com/dashpay/dash/issues/2198 but doesn't remove the 10 denom. I'd still like to see that change done. I don't think the 10 has enough usage for it to retain its privacy

PastaPastaPasta commented:
Yesterday there were only 24 mixes using 10 denoms. I would consider that very very low mixing volume

UdjinM6 commented:
Low number of mixing sessions - yes, but low volume - I wouldn't say so.

Full stats for yesterday are like this:
24x10 vs 80x1 vs 185x0.1 vs 312x0.01

For 1 input per session that would be:
240 vs 80 vs 18.5 vs 3.12
(per participant)

Yes, avg number of inputs might be higher for smaller denoms. Let's say it's 5 inputs for all other denoms but 10s:
240 vs 400 vs 92.5 vs 15.6
(per participant)

Which still gives the 2nd place by volume for 10s and it would be the 1st place if avg for 10s would be 2 inputs per session (per participant) instead.

PastaPastaPasta commented:
I don't see how that is relevant to its effective privacy. Low volume was likely bad word choice on my side as the number of mixing sessions is what really matters

UdjinM6 commented:
The problem with removing 10s is basically like this: if you are mixing large amounts and want to spend > 10 DASH via PS later you need 10s because otherwise you are stuck with so many 1s in 1 spending tx that it becomes not-so-private for cluster analysis algos. Having not so many sessions is not _that_ bad imo.

PastaPastaPasta commented:
Agreed, lack of sessions concerns me but I agree with you on the cluster analysis remark

gladcow commented:
utACK

nmarley commented on src/privatesend.h:
```diff
@@ -409,8 +397,8 @@ class CPrivateSend
     static bool IsCollateralValid(const CTransaction& txCollateral);
-    static CAmount GetCollateralAmount() { return COLLATERAL; }
-    static CAmount GetMaxCollateralAmount() { return COLLATERAL*4; }
+    static CAmount GetCollateralAmount() { return vecStandardDenominations.back()/10; }
```
Would prefer to not assume knowledge of `vecStandardDenominations` here and instead use some kind of `GetSmallestDenomination`, but I guess we can do some clean up later.

(Only because this exposes internals of how we manage PS denoms, and we are directly accessing the internal vector instead of using accessors, which provide an abstraction and allow calling code to not worry about implementation.)

nmarley commented on src/privatesend.h:
```diff
@@ -204,11 +199,6 @@ class CDarksendQueue
     template <typename Stream, typename Operation>
     inline void SerializationOp(Stream& s, Operation ser_action) {
         READWRITE(nDenom);
-        int nVersion = s.GetVersion();
```
@UdjinM6 Will this change the wire serialization of DSQ (and cause incompatibility between versions)? Or does that not matter since Spork 6 was never activated?

UdjinM6 replied:
Yes, serialization will change, but we are bumping `MIN_PRIVATESEND_PEER_PROTO_VERSION` too here (and we don't process PS messages from/to to nodes below it), so it should be fine.

InhumanPerfection replied:
All 12.4 nodes will reject DSA and DSQ from any 12.3 node, and all 12.3 nodes will not be able to decode DSA and DSQ from any 12.4 node.

I see 2 issues here during the network upgrade:
1) there may be problems with propagation of DSQ messages through the network,
2) 12.3 nodes can mix only with another 12.3 nodes only through 12.3 MNs, the same for 12.4 nodes/MNs

The Second issue would not be so big if the First issue didn't make mixing unstable for quite long time. Thoughts?

UdjinM6 replied:
Small correction: 12.4 won't relay any PS message to 12.3 nodes. But yes, the migration issues you highlighted are correct - mixing usability will be degraded for a week or so while network is upgrading, especially if you are not using the version which is used by the majority (MNs/mixing participants) atm.

UdjinM6 commented:
Rebased to fix merge conflict (due to dropping `nInputCount`). Also addressed the comment regarding using `GetSmallestDenomination` while at it. Another thing I have to note (after reviewing old PRs) is that we'll probably need smth like #1372 now that the smallest denom is lower.

InhumanPerfection commented:
I'm really don't like transition process because of [this](https://github.com/dashpay/dash/pull/2318#discussion_r225471436). But this is one time process, so f it? ¯\_(ツ)_/¯

UdjinM6 commented:
Bumped `MIN_PRIVATESEND_PEER_PROTO_VERSION` to the most recent proto version (70212).

UdjinM6 commented:
Rebased to fix merge conflicts after #2373 . Ready for the final review :)

UdjinM6 commented:
NOTE: revert this ( https://github.com/dashpay/dash/commit/262454791c4b4f783b2533d1b16b757a71eb5f7d ) is you want to keep mixing on current testnet/mainnet
